Wire order conversion

LaurinFischer opened the pull request:
### Summary
This PR implements the conversion between the wire orders "sequential" and "interleaved". 
Firstly, the circuit payload that is sent to the backend is converted so that the qire indices of the instructions match the expectation of the backend. 
Secondly, the result strings obtained from the backend are converted back to Qsikit convention if necessary. 

This way, for the Qiskit user, everything will always be consistent with the "sequential" wire order. 

### Details and comments
- backends that use multiple atomic or spin species must specify so in their configuration dict through a "num_species" key. 
This needs to implemented on Fred's side
- Still needs tests and probably some debugging
- When fully working, we can adjust the tutorials so that the same circuits are sent to both local and remote backends. also postprocessing of the circuits should then be identical.

eggerdj reviewed (changes requested):
I think this looks like what we want to have. Can you add a couple of unit tests for the reordering logic?

fretchen commented:
Hi,

I start to think that this involves more moving parts than initially expected.

- The whole configuration is also assuming the `wire_order` and should be handled consistently.  Take the hopping gate for example. The whole `gate_map` has to be adapted too as it tells the user which gates should be connected etc. 
- Now let us assume that we get the whole thing working properly. Any kind of debugging between the backend and the uneducated user will be really hard as everything has to be transformed back and forth.
- Take the case of some not very careful back-end developer that does not know what `wire_order` means. They will all have a super harsh time to figure out what is going wrong.
- I think that the naming `num_species` is to close to `atomic_species`. It will be very hard to know intuitively what is meant for each one.

So, I actually start to think that this might be better fit in transpiler pass and that we should let the user know the default which kind of ordering the back-end assumes.

LaurinFischer commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+        convention_to: str,
+        num_sites: int,
+        num_species: int,
+        sort: Optional[bool] = False,
```
A sequence of indices in one wire convention might not be sorted in the other convention. E.g. `[0, 1, 4, 5]` in "sequential" (4 sites, two spins) goes to `[0, 2, 1, 3]` in "interleafed". I wanted some flexibility because I believe that for some purposes this should be ordered (e.g. checking against the backend coupling map) and for others not.

LaurinFischer commented on qiskit_cold_atom/providers/cold_atom_job.py:
```diff
@@ -88,12 +94,72 @@ def result(self, timeout: Optional[float] = None, wait: float = 5.0) -> Result:
         result_dict = self._wait_for_result(timeout, wait=wait)
 
-        return Result.from_dict(result_dict)
+        if "num_species" in self._backend.configuration().to_dict():
```
thanks, see 2684a9f

LaurinFischer commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+        if convention_from == convention_to:
+            new_wires = wires
+
+        if convention_from == "sequential" and convention_to == "interleaved":
+            new_wires = [idx % num_sites * num_species + idx // num_sites for idx in wires]
+
+        elif convention_from == "interleaved" and convention_to == "sequential":
+            new_wires = [idx % num_species * num_sites + idx // num_species for idx in wires]
```
see 2684a9f

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -99,8 +101,15 @@ def validate_circuits(
+            if "num_species" in config_dict:
                 num_species = backend.configuration().num_species
+                if "wire_order" in config_dict:
+                    wire_order = backend.configuration().wire_order
```
```suggestion
                    wire_order = WireOrder(backend.configuration().wire_order)
```

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -153,9 +174,28 @@ def circuit_to_data(cls, circuit: QuantumCircuit) -> List[List]:
+        if "num_species" in config_dict:
+            num_species = backend.configuration().num_species
+            if "wire_order" in config_dict:
+                wire_order = backend.configuration().wire_order
```
```suggestion
                wire_order = WireOrder(backend.configuration().wire_order)
```

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
         cls.validate_circuits(circuits=circuits, backend=backend, shots=shots)
 
+        if "wire_order" in backend.configuration().to_dict():
+            wire_order = backend.configuration().wire_order
```
```suggestion
            wire_order = WireOrder(backend.configuration().wire_order)
```

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+    def convert_wire_order(
+        cls,
+        wires: List[int],
+        convention_from: str,
+        convention_to: str,
```
```suggestion
        convention_from: WireOrder,
        convention_to: WireOrder,
```

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+        Returns:
+            A list of wire indices following the convention_to.
+        """
+        if (convention_to or convention_from) not in ["sequential", "interleaved"]:
```
```suggestion
        if (convention_to or convention_from) not in WireOrder:
```
However, this check may no be needed anymore if we assume that instances of `WireOrder` are given.

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+        if convention_from == convention_to:
+            new_wires = wires
+
+        if convention_from == "sequential" and convention_to == "interleaved":
```
```suggestion
        if convention_from == WireOrder.SEQUENTIAL and convention_to == WireOrder.INTERLEAVED:
```

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -185,13 +225,71 @@ def circuit_to_cold_atom(
+        if convention_from == "sequential" and convention_to == "interleaved":
+            new_wires = [idx % num_sites * num_species + idx // num_sites for idx in wires]
+
+        elif convention_from == "interleaved" and convention_to == "sequential":
```
```suggestion
        elif convention_from == WireOrder.INTERLEAVED and convention_to == WireOrder.SEQUENTIAL:
```

eggerdj commented on qiskit_cold_atom/providers/cold_atom_job.py:
```diff
@@ -88,12 +94,73 @@ def result(self, timeout: Optional[float] = None, wait: float = 5.0) -> Result:
+
+        num_species = self._backend.configuration().num_species
+        try:
+            wire_order = self._backend.configuration().wire_order
```
```suggestion
            wire_order = WireOrder(self._backend.configuration().wire_order)
```

eggerdj commented on test/test_circuit_tools.py:
```diff
@@ -203,6 +203,30 @@ def test_circuit_to_data(self):
+
+        with self.subTest("test sequential to interleaved"):
+            wires_converted = CircuitTools.convert_wire_order(wires=wires_sequential,
+                                                              convention_from="sequential",
+                                                              convention_to="interleaved",
```
```suggestion
                                                              convention_from=WireOrder.SEQUENTIAL,
                                                              convention_to=WireOrder.INTERLEAVED,
```

eggerdj commented on test/test_circuit_tools.py:
```diff
@@ -203,6 +203,30 @@ def test_circuit_to_data(self):
+
+        with self.subTest("test interleaved to sequential"):
+            wires_converted = CircuitTools.convert_wire_order(wires=wires_interleaved,
+                                                              convention_from="interleaved",
+                                                              convention_to="sequential",
```
```suggestion
                                                              convention_from=WireOrder.INTERLEAVED,
                                                              convention_to=WireOrder.SEQUENTIAL,
```

eggerdj reviewed (commented):
I think we can make the convention a bit more prominent. To do this add the following in circuit_tools.py:
```python
from enum import Enum

class WireOrder(str, Enum):
    """The possible wire orderings for cold atomic circuits.
    
    For example, a sequential register [0, 1, 2, 3, 4, 5] with two species implies that wires 0, 1, 2
    are of the same type while an interleaved ordering implies that wires 0, 2, and 4 are of the
    same type.
    """
    SEQUENTIAL = "sequential"
    INTERLEAVED = "interleaved"
```
this now means that you can do:
```python
wire_order = WireOrder("sequential")
```
note that the following will raise an error:
```python
wire_order = WireOrder("something random")
```
which gives some extra safety.

eggerdj commented on test/test_circuit_tools.py:
```diff
@@ -206,3 +206,27 @@ def test_circuit_to_data(self):
+
+        num_sites = 4
+        num_species = 3
+        wires_sequential = [0, 1, 4, 5, 8, 9]
+        wires_interleaved = [0, 3, 1, 4, 2, 5]
```
This is a good example that this ordering is not a trivial issue. So essentially, position 1 in sequential becomes position 3 in interleaved because we have three species? Is it possible to write down the conversion rule as a one-lines comment?

LaurinFischer replied:
Added `# conversion rule: i -> (i % num_sites) * num_species + i // num_sites`

LaurinFischer commented:
I have now implemented the discussed changes:

- Conversion of wires to the backend convention now works seemlessly. This conversion affects only the formatting of the JSON payload that is sent to the backend.
- I have finally figured out a convenient way to avoid having to refactor any measurement results that are communicated back by the backend. This is done by changing the order of the measurement instructions in the submitted Json payload. 
- The fermionic tweezer tutorial is now updated to reflect these changes. Note how in the part that interacts with the remote simulator, the circuits are now also built with the standard `sequential` ordering. The processing of results is done with a new helper routine `mean_counts_right()` which can be applied to both the remote and local jobs. This exemplifies that the processing of results now also works the same for both backend wiring conventions (even when only some of the wires in the register are measured)
- As Fred requested, there is also the option now to turn off the wire ordering with a boolean flag `convert_wires` in the `backend.run()` method. This defaults to `True`. If manually set to `False`, circuits with interleaved wire ordering can be sent without having Qiskit interfere.

LaurinFischer commented on qiskit_cold_atom/providers/cold_atom_job.py:
```diff
@@ -88,12 +94,72 @@ def result(self, timeout: Optional[float] = None, wait: float = 5.0) -> Result:
+
+            reordered_result_dict = deepcopy(result_dict)
+
+            for exp_idx, circ_result in enumerate(reordered_result_dict["results"]):
```
I believe that this part is no longer required. Getting the order of the measurement results right is now done by adjusting the order of the measure instructions in the circuit JSON payload. I would argue that this is the much more natural and flexible way. For example, 
```     
...
['measure', [0], []],
['measure', [1], []],
['measure', [4], []],
['measure', [5], []],
...
```
is converted to 
```     
...
['measure', [0], []],
['measure', [2], []],
['measure', [1], []],
['measure', [3], []],
...
```
It's the order of these measurement instructions that defines the order of the entries in the result bitstring.
Hence, by manipulating the order of these instructions in the right way, the resulting bitstrings already have the correct ordering that we need.

fretchen commented:
Looks cool from the user perspective. In the notebook, you still have to update the part that saves the credentials too:

```
# save an account to disk
provider = ColdAtomProvider.save_account(
    url=[
        "http://qsimsim.synqs.org/api/singlequdit",
        "http://qsimsim.synqs.org/api/multiqudit",
        "http://qsimsim.synqs.org/api/fermions",
    ],
    username="name",
    token="password",
)
```

The `/api/` is missing for the moment

eggerdj commented on qiskit_cold_atom/circuit_tools.py:
```diff
@@ -100,21 +115,20 @@ def validate_circuits(
 
             # If num_species is specified by the backend, the wires describe different atomic species
-            # and the circuit must exactly match the expected wire count of the backend.
+            # and the circuit's wire count must be a multiple of this.
```
```suggestion
            # and the circuit's wire count must be a multiple of the number of species.
```

eggerdj commented on qiskit_cold_atom/providers/remote_backend.py:
```diff
@@ -92,6 +93,7 @@ def run(
         Args:
             circuit: The quantum circuits to be executed on the device backend
             shots: The number of measurement shots to be measured for each given circuit
+            convert_wires: If True, the circuits are converted to the wiring convention of the backend.
```
```suggestion
            convert_wires: If True (the default), the circuits are converted to the wiring convention of the backend.
```

eggerdj reviewed (changes requested):
This looks good to me. Before approval please merge in the suggestions. I also noted the following in the NB:
```python
# or enable an account in the current session
provider = ColdAtomProvider.enable_account(
    url=[
        "http://qsimsim.synqs.org/api/singlequdit",
        "http://qsimsim.synqs.org/api/multiqudit",
        "https://qsimsim.synqs.org/api/fermions",
    ],
    username="name",
    token="password",
)
```
Can we use `https` instead of `http`? If so please change to `https`.

eggerdj reviewed (approved):
LGTM. Thanks
